From 45b68dcd4087a6f220bb51d31534cbe96425a685 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sat, 15 Feb 2020 15:42:26 +0100
Subject: [PATCH] arm64: irq_pipeline: skip n/a lockdep check in kernel
 preemption

When irqs are pipelined, the root stage must be unstalled when
branching to preempt_schedule_irq(), although hard irqs must be off.
Skip the lockdep check in the asm trampoline to preempt_schedule_irq()
when interrupts are pipelined. A series of debug checks will be
applied by the latter, including this one.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 arch/arm64/kernel/process.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/kernel/process.c b/arch/arm64/kernel/process.c
index 393ca414d0e9..455f4c15d7ff 100644
--- a/arch/arm64/kernel/process.c
+++ b/arch/arm64/kernel/process.c
@@ -636,7 +636,8 @@ core_initcall(tagged_addr_init);
 
 asmlinkage void __sched arm64_preempt_schedule_irq(void)
 {
-	lockdep_assert_irqs_disabled();
+	if (!irqs_pipelined())
+		lockdep_assert_irqs_disabled();
 
 	/*
 	 * Preempting a task from an IRQ means we leave copies of PSTATE
-- 
2.16.4

